### DISCUSSION TITLE
Cloudwatch plugin - allow expressions to do division

### DISCUSSION THREAD

**Discussion by rgrullon** [grafana/grafana#43142] [Plugins]
This is a feature request to allow for expressions to do division as they seem to be able to do in Cloudwatch.

Goal is to be able to both graph and create an alert for error rate. This is with using grafana 8.1.2.

In Cloudwatch - you can do an expression like m1+m2/m3.
The expression works and data is shown. You do get a warning back about " One or more data-points have been dropped due to non-numeric values (NaN, -Infinite, +Infinite)"
(see snip1)

In grafana, I setup the same queries with the Cloudwatch datasource. The queries work but itself, but not the Expression/"metric math". Receives a similar warning in the grafana logs

" t=2021-09-10T11:49:03+0000 lvl=eror msg="Metric request error" logger=context userId=68 orgId=1 uname="Trovato, Philip" error="failed to query data: ArithmeticError in query \"D\": One or more data-points have been dropped due to non-numeric values (NaN, -Infinite, +Infinite)" remote_addr=172.23.78.64"

But unlike Cloudwatch, grafana does not return any data when this warning occurs. Just the "metric request error". I can do other expressions fine, like m1 + m2. But really need to be able to divide and do an error rate.
(see snip2)

Goal is to be able to both graph and create an alert for error rate. This is with using grafana 8.1.2.

https://camo.githubusercontent.com/dfcfbad700bbbf7df7062286713dcda1ac422e669ffac74d4fd398b8a4800f7b/68747470733a2f2f67726166616e612e7a656e6465736b2e636f6d2f6174746163686d656e74732f746f6b656e2f5453393668773767736652474d6e636b4465715a494d49546f2f3f6e616d653d43617074757265352e504e47

**Comment by belaaiza** [comment]
As a workaround, I was able to this by doing m1+m2/SUM(m3), maybe this works for you two

**Reply by rhowe-trp** [reply]
Using `m1+m2/SUM(m3)` is not the same maths as `m1+m2/m3`

I think the Grafana Cloudwatch datasource needs to behave the same as Cloudwatch here and ignore the warning from cloudwatch about dropped datapoints due to divide by zero as described here: https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/using-metric-math.html#metric-math-syntax-arithmetic

**Comment by pandar00** [comment]
It looks like there's a difference in how CloudWatch renders the chart on the console vs Grafana.

In the case of ArithmeticError, CloudWatch tries to render the chart with whatever data is available in the response whereas Grafana rejects it completely. This is possible because the query resulting in an Arithmetic error has a response code of 400 but has some metric values.

Example response http://jsonblob.com/971919418250772480. Metric "D" resulted in an error but has some data points available to plot

An option to override this behavior in Grafana would be great

**Comment by kumartushar** [comment]
I am facing the same issue with Grafana and Cloudwatch.

**Comment by ionstormadi** [comment]
Same issue while trying to get EFS throughput utilization in percentage

**Comment by HenryNguyen5** [comment]
Would one of these work as a stop-gap?
- `IF((errors / invocations) >= 0, errors / invocations, 0)` -> If errors / invocations produces a valid number, since we should always have 0 as the floor, then return that value, otherwise default to 0
-  `errors / FILL(invocations, 1)` -> Make sure there is always an invocation data point to divide by

**Comment by sarahzinger** [comment]
Hey all, sorry for the delay in looking at this. Trying to determine if this is still an issue at this point or if this request has already been addressed.

I just gave this a shot on a more recent version of grafana and was not able to reproduce the issue so I'm guessing it has been resolved, but if you all are still running into this please feel free to open an issue in the main repo!

Thanks so much!

**Reply by polak785** [reply]
Hello,

Yes I'm facing the problem with error `ArithmeticError in query \"F\": One or more data-points have been dropped due to non-numeric values (NaN, -Infinite, +Infinite)` 

Sometime it work, sometime I have this error. The errors occurs more often unfortunately.

**Reply by sarahzinger** [reply]
Interesting I am able to get an Arithmetic Error back if I divide by zero, is it possible that's happening in some of your queries, which would explain why it sometimes works and sometimes doesn't? If you're getting a discrepancy in data with the same query from Grafana and the AWS Console, if you can open a ticket with the details that would be very helpful!

I also noticed that our error handling could be improved for Cloudwatch Metric Search Queries and have made a ticket for that here https://github.com/grafana/grafana/issues/78819

**Reply by polak785** [reply]
@sarahzinger I create it last time :) #78608
